pkg/querier/tail.go

```diff
@@ -0,0 +1,92 @@
+package querier
+
+import (
+	"context"
+	"time"
+
+	"github.com/grafana/loki/pkg/iter"
+	"github.com/grafana/loki/pkg/logproto"
+)
+
+// tailIteratorIncrement is for defining size of time window we want to query entries for
+const tailIteratorIncrement = 10 * time.Second
+
+// delayQuerying is for delaying querying of logs for specified seconds to not miss any late entries
+const delayQuerying = 10 * time.Second
+
+func (q *Querier) tailQuery(ctx context.Context, queryStr, regexp string) iter.EntryIterator {
+	return &tailIterator{
+		from:     time.Now().Add(-(tailIteratorIncrement + delayQuerying)),
+		queryStr: queryStr,
+		regexp:   regexp,
+		querier:  q,
+		ctx:      ctx,
+	}
+}
+
+type tailIterator struct {
+	queryStr string
+	regexp   string
+	from     time.Time
+	err      error
+	ctx      context.Context
+	querier  *Querier
+	iter.EntryIterator
+}
+
+func (t *tailIterator) Next() bool {
+	var err error
+	for t.EntryIterator == nil || !t.EntryIterator.Next() {
+		through, now := t.from.Add(tailIteratorIncrement), time.Now()
+		if through.After(now.Add(-delayQuerying)) {
+			time.Sleep(through.Sub(now.Add(-delayQuerying)))
+		}
+
+		t.EntryIterator, err = t.query(t.from, through)
+		if err != nil {
+			t.err = err
+			return false
+		}
+
+		// We store the through time such that if we don't see any entries, we will
+		// still make forward progress. This is overwritten by any entries we might
+		// see to ensure pagination works.
+		t.from = through
+	}
+
+	return true
+}
+
+func (t *tailIterator) Entry() logproto.Entry {
+	entry := t.EntryIterator.Entry()
+	t.from = entry.Timestamp.Add(1 * time.Nanosecond)
+	return entry
+}
+
+func (t *tailIterator) Error() error {
+	return t.err
+}
+
+func (t *tailIterator) query(from, through time.Time) (iter.EntryIterator, error) {
+	request := &logproto.QueryRequest{
+		Query:     t.queryStr,
+		Limit:     30,
+		Start:     from,
+		End:       through,
+		Direction: logproto.FORWARD,
+		Regex:     t.regexp,
+	}
```

Comment:
I'd potentially store the original QueryRequest in the tailIterator struct, use proto.Clone to get a copy on each query, and update the from and through in the copy.  That way when we add new fields we don't need to remember to copy them here.

Reply:
Actually I am building QueryRequest because we receive request in query params, same as GET request handler here:
https://github.com/grafana/loki/blob/f2254d667ef4ae7674e4b9f5e666c4f0f037972b/pkg/querier/http.go#L91-L98

So there is no original QueryRequest I guess. Did I miss something?

Reply:
Ah good point.  Can you try and reuse the queryrequest parsing from the query?